README.md
# VKGroupLikes - VK Group Likes Bot #
## По-русски
`VK Group Likes Bot` - бот для автоматизации работы с группами Vk<br/>
Т.е, пока что ты можешь автоматизировать:
  - Автоматически ставить лайки на новые посты в группе
  - Пролайкать последние n-постов в группе
## Как начать ?
Ну что решились? Что ж начнем ...
### Установка Python
Для начала нам потребуется скачать [Python](https://www.python.org/downloads/)  
Устанавливаем и не забываем на всякий случай поставить галочку в __add to path__  
Установили? Молодцы!
### Теперь Настраеваем
Для начала скачиваем и кидаем в какую-нибудь папку
#### В результаате у нас должно получиться 2 файла:  
  - main.py - Главный Файл
  - config.py - Конфиг (Его то мы и будем редактировать)  
## Разберём Конфиг
Открыть `config.py` можно через обычный блокнот
  - 'login' и 'password' - Логин и Пароль соответственно  
  - 'token' - Токен (Об этом позже)  
  - 'group_id' - id Группы (Имеет вид -xxxxxxxxx)  
  - 'count_last_posts' - Кол-во последних постов, на которые нужно поставить лайки  
  - 'liked_last_posts' - Лайкать ли последние посты группы (Значение True или False)
  - 'liked_new_posts' - Лайкать ли новые посты группы (Значение True или False)
  - 'cooldown_like' - Задержка в секундах между постановкой следующего лайка (По умолчанию: 0.4 сек)  
  - 'cooldown_check' - Частота проверки новых постов в группе в секундах (По умолчанию 600 сек = 10 минут)  
## Доступ к аккаунту:  
  * Через __Логин/Пароль__ пишем в поля YOUR_LOGIN/YOUR_PASSWORD cвой Логин/Пароль соответственно  
  * Через __Токен__, переходим по ссылке [https://oauth.vk.com/token?grant_type=password&client_id=2274003&client_secret=hHbZxrka2uZ6jB1inYsH&username=**login**&password=**password**](https://oauth.vk.com/token?grant_type=password&client_id=2274003&client_secret=hHbZxrka2uZ6jB1inYsH&username=**login**&password=**password**)  
Где **login** и **password**, логин и пароль соответственно
"access_token":"`adsafbc78772134115301f197ece048092314fe3db2334bb3852ec4b025df169853421360fd1asfd4`"  
Копируем всё что в кавычках после __access_token__ и вставляем в поле YOUR_TOKEN  
## Где найти id группы?
Открываем любую группу и открываем любой пост  
https://vk.com/{Имя_Группы}?w=wall`-1XXXXXXXX`_`XXX`, где
* -1XXXXXXXX - ID Группы
* XXX - ID Поста
## RoadMap
- [x] Авто установка модуля Vk-Api
- [ ] Пофиксить (Не лакает последний пост, если он закреплен)
- [ ] Добавить обход капчи (Если лайкать 50 и более постов подряд, то вк её запросит)
- [ ] Добавить цвета в консоль


CHANGELOG.md
## [0.51] - (19-Apr-2019) ##  
### Fixed  
  - Auto-install vk_api module  
  - Some other fixes  
## [0.50] - (18-Apr-2019)  
  - Upload to GitHub  


main.py
# -*- coding: utf-8 -*-
#   Import standart libraries and vk_api
import requests, time, config, datetime
#   Install vk_api from pip
from os import system
try:
    import vk_api
except ImportError:
    print('Installing VK_API ...')
    system('pip install vk_api')
    import vk_api
#   Import config data
config = config.config
__version__ = 0.2
last_post_from_group = 0
print('Starting VKGroupLikes ...')
#   function start VK session
def vk_login():
    if config.get('token') != ('' or 'YOUR_TOKEN'):
        print('Successful login with token')
        vk_session = vk_api.VkApi(token=config.get('token'))
    elif (config.get('login') != ('' or 'YOUR_LOGIN') and config.get('password') != ('' or 'YOUR_PASSWORD')):
        print('Successful login')
        vk_session = vk_api.VkApi(config.get('login'), config.get('password') )
        vk_session.auth()
    elif config.get('group_id') == ('' or 'YOUR_GROUP_ID'):
        print('Please set group id')
        exit()
    else:
        print("You have not provided a login and password or token.\nLogin like: email or +7xxxxxxxxxx")
        exit()
    vk = vk_session.get_api()
    return vk
#   function check was there a post have like(return True or False)
def is_liked(post_id):
    if (vk.likes.isLiked(type = 'post', owner_id = config.get('group_id'), item_id = post_id)['liked']) == True:
        return True
    else:
        return False
#   function check last id(-1) post from group. (-1)id Because group may be have fixed post. return number
def get_last_id():
    last_id = vk.wall.get(owner_id = config.get('group_id'), count = 10)['items'][1]['id']
    return last_id
#   function return str current time
def get_time():
    today = datetime.datetime.today()
    return today.strftime("%H:%M:%S")
#   function liked last n posts
def last_posts(last_post, count):
    start_post = last_post - count
    last_post += 1; start_post += 1;
    time_start = time.time()
    print('{} Last Post is {}'.format(get_time(), str(get_last_id())) )
    for id in range(start_post, last_post):

            if is_liked(id) == False:
                try:
                    vk.likes.add(type = 'post', owner_id = config.get('group_id'), item_id = id)
                    print('{} Liked post {}'.format(get_time(), str(id)) )
                except vk_api.exceptions.ApiError:
                    print('{} Post Deleted {}'.format(get_time(), str(id)) )
                except vk_api.exceptions.ApiHttpError:
                    print( get_time() + ' Network Error')
            else:
                print(get_time() + ' Post Already Liked ' + str(id))
            time.sleep(config.get('cooldown_like'))
            result_time = time_start - time.time()
    print('{} All Done in {} seconds'.format(get_time(), str(abs(round(result_time, 3)))) )
    global last_post_from_group; last_post_from_group = id
#   function Liked all new posts
def new_posts():
    global last_post_from_group
    if last_post_from_group == 0:
        last_post_from_group = get_last_id()
    while True:
        if get_last_id() != last_post_from_group and not is_liked(get_last_id()):
            try:
                vk.likes.add(type = 'post', owner_id = config.get('group_id'), item_id = get_last_id())
                print( get_time() + ' Liked New Post!')
                last_post_from_group = get_last_id()
            except vk_api.exceptions.ApiError:
                print('{} Post Deleted {}'.format(get_time(), str(get_last_id())) )
                last_post_from_group = get_last_id()
            except vk_api.exceptions.ApiHttpError:
                print( get_time() + 'Network Error')
        else:
            print('{} All new posts liked, waiting {} seconds'.format(get_time(), config.get('cooldown_check')))
            time.sleep(config.get('cooldown_check'))
# Start Script and select mode
if __name__ == "__main__":
    vk = (vk_login())
    if config.get('liked_last_posts'):
        last_posts(get_last_id(), config.get('count_last_posts'))
    else:
        print('Edit config and select between liked_last_posts or liked_last_posts\nP.S Possible choose both')
    if config.get('liked_last_posts'):
        new_posts()
    else:
        print('Edit config and select between liked_last_posts or liked_last_posts\nP.S Possible choose both')


config.py
# -*- coding: utf-8 -*-

config = {
    'login'             :   'YOUR_LOGIN',       # Your VK login
    'password'          :   'YOUR_PASSWORD',    # Your VK password
    'token'             :   'YOUR_TOKEN',       # Your VK token
    'group_id'          :   '-1XXXXXXXX',		# Group id Like -1XXXXXXXX without ''
    'count_last_posts'  :    10,	            # Count of last posts to be 				   | 10 posts
    'liked_last_posts'  :    True,			    # Procedure of Liked Last Count of Wall Posts. | True or False
    'liked_new_posts'   :    True,		        # Start deamon to liked new group post.        | True or False
    'cooldown_like'     :    0.4,               # Cooldown between liked last group posts in seconds. !Very low cooldown may cause captcha error
    'cooldown_check'    :    600                # Cooldown between check new posts from group in seconds.
}
